### DISCUSSION TITLE
Using .Net MAUI

### DISCUSSION THREAD

**Discussion by MaroisConsulting** [dotnet/maui#5253] [General]
I have created apps with Xamarin Forms and WPF. I would create a WPF project with the Windows Views in it, and a Xamarin Forms project with the Mobile Views in it, and a common project with items shared by both projects. See the image.

The question I have about MAUI, is how to structure things so I can still create both Windows and Android apps. Can someone point me to an example of this?

Thanks
![Paragon Web API](https://user-images.githubusercontent.com/1486240/157993133-26b6f817-f1a8-4176-83ab-fad0eee14a10.png)

**Comment by nogginbox** [comment]
I think it depends on how much you want to share and what's different about the bits you don't.

I'm currently sharing everything, but when I get to the testing stage I expect I'm going to use some onplatform stuff to switch a few views around. Or I May do this based on screen size, or device type. This is because I think the UI will need to change based on how the user is using it.
 
I'm not planning on using any platform specific APIs, but if I was, then I would read the [MAUI docs on invoking platform code](https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/invoke-platform-code) and the [platform specific code section of the single project docs](https://docs.microsoft.com/en-us/dotnet/maui/fundamentals/single-project#platform-specific-code).

**Reply by MaroisConsulting** [reply]
OK, thanks. But would you create seperate MAUI projects for each target Win & Android?

**Reply by matt-goldman** [reply]
> OK, thanks. But would you create seperate MAUI projects for each target Win & Android?

I think that depends on whether you are making the same app, but for different platforms, or making fundamentally different apps. If they are the same app, I would make these in a single .NET MAUI project, and create different views for different platforms if that's the goal, or just customise their appearance and behaviour depending on platform or idiom. If they are different apps but accessing the same back end and sharing some logic, I would do it the same way you are doing now. Create a solution that contains the back end and individual .NET MAUI projects. And of course a class library containing the shared logic.

**Reply by MaroisConsulting** [reply]
The app is a construction managment app in WPF. Now my client wants to have some mobile parts for their foremen. The ViewModels and other parts can be shared. 

When you say "create different views for different platforms if that's the goal, or just customise their appearance and behaviour depending on platform or idiom." how does the app know where its running?

**Reply by matt-goldman** [reply]
There are three ways you can handle this for. The first is with a compiler directive that only builds certain things per platform. To see how this works, have a look at how David Ortinau does it in his WeatherTwentyOne sample app: https://github.com/davidortinau/WeatherTwentyOne/blob/main/src/WeatherTwentyOne/MauiProgram.cs. The other way is to use the Essentials library. This can give you information about platform (Windows, macOS, iOS, or Android) or idiom (desktop, mobile or platform) at runtime. You can also use markup extensions in XAML to control how things are displayed either based on idiom or platform.

**Comment by Xyncgas** [comment]
One unified front end is the selling point of web technology, so I just build UI in a razor class library and simply use it in MAUI project, then it runs perfectly in any platform with one UI and mobile screen detection and platform detection (DIY)

**Reply by Xyncgas** [reply]
There is just one MAUI project for all platform, and all the other projects for as many as you want to have to make the app working
